### PR TITLE
feat: support undefined terms in the input document

### PR DESCRIPTION
**Pull request by zhenwenc**
This PR enhances the library to support encoding JSON-LD documents and decoding CBOR-LD documents with compression enabled while doesn't require all terms in the input document to be defined in the declared `@context`.

### Scope

- The changes in this PR will be backward compatible, the undefined terms support can be enabled by setting `compressionModeUndefinedTermAllowed` to `true` (defaults to `false`).
- Affects the encode/decode functionality only when compression is enabled.

### Risk

- This will affect the compression rate of the CBOR-LD encoded document when the source JSON-LD document contains undefined terms. In the worst-case scenario, if all the terms are undefined, the result will be more or less the same as encoding without compression.
- The CBOR-LD encoded document will not be able to be decoded once any dependent context had been modified, which was already a known limitation of this library.

### Description

There are cases that a JSON-LD document contains terms that are not defined in any of the declared contexts, based on the JSON-LD specification, these terms will be expanded with the `@vocab` [default vocabulary](https://www.w3.org/TR/json-ld11/#default-vocabulary), which isn't supported by this library.

Instead of failing the encode/decode operations will `ERR_UNKNOWN_CBORLD_TERM_ID` error, we allow optionally skip compressing the unknown terms by preserving the original term key in the transform map, which can be controlled by the `compressionModeUndefinedTermAllowed: boolean` option in both `encode` and `decode` functions (defaults to `false`).

This will enable us to encode JSON-LD documents with compression like this example, where the "proof.proofValue" term hasn't been defined anywhere: https://github.com/w3c/vc-test-suite/blob/cfad6f390316529e32ef0a42d1330e16ba42968c/test/vc-data-model-1.0/input/example-015-zkp.jsonld

Erro with `compressionModeUndefinedTermAllowed: false`:
```
Error: ERR_UNKNOWN_CBORLD_TERM: Unknown term 'proofValue' was detected in the JSON-LD input.
 at new CborldError (lib/CborldError.js:9:19)
 at Compressor._getEntries (lib/Compressor.js:134:15)
 at Compressor._transform (lib/Transformer.js:56:30)
 at processTicksAndRejections (internal/process/task_queues.js:97:5)
 at async Compressor._transformObject (lib/Compressor.js:191:5)
 at async Compressor._transform (lib/Transformer.js:122:9)
 at async Compressor._createTransformMaps (lib/Compressor.js:90:7)
 at async Compressor.compress (lib/Compressor.js:70:27)
 at async encode (lib/encode.js:53:14)
```

When encoding with `compressionModeUndefinedTermAllowed: true`, the following transform map will be produced after compression:
```
Diagnostic CBOR-LD compression transform map(s):
Map {
  1 => [
    ContextEncoder {
      context: 'https://www.w3.org/2018/credentials/v1',
      appContextMap: Map {}
    },
    ContextEncoder {
      context: 'https://www.w3.org/2018/credentials/examples/v1',
      appContextMap: Map {}
    }
  ],
  114 => Map {
    'proofValue' => 'DgYdYMUYHURJLD7xdnWRinqWCEY5u5fK...j915Lt3hMzLHoPiPQ9sSVfRrs1D',
    116 => 'ex:AnonCredPresentationProofv1'
  },
  116 => VocabTermEncoder { termId: 110 },
  542 => Map {
    1 => [
      ContextEncoder {
        context: 'https://www.w3.org/2018/credentials/v1',
        appContextMap: Map {}
      },
      ContextEncoder {
        context: 'https://www.w3.org/2018/credentials/examples/v1',
        appContextMap: Map {}
      }
    ],
    544 => Map {
      112 => 'did:example:cdf:35LB7w9ueWbagPL94T9bMLtyXDj9pX5o',
      116 => 'did:example:schema:22KpkXgecryx9k7N6XN1QoN3gXwBkSU8SfyyYQG'
    },
    548 => Map { 500 => 'College of Engineering', 502 => 'BachelorDegree' },
    554 => XsdDateTimeEncoder {
      value: '2010-01-01T19:23:24Z',
      parsed: 1262373804000
    },
    558 => 'did:example:Wz4eUg7SetGfaUVCn8U9d62oDYrUJLuUtcy619',
    114 => Map {
      518 => 'mu6fg24MfJPU1HvSXsf3ybzKARib4WxG...RSce53M6UwQCxYshCuS3d2h',
      522 => 'cg7wLNSi48K5qNyAVMwdYqVHSMv1Ur8i...Fg2ZvWF6zGvcSAsym2sgSk737',
      116 => 'ex:AnonCredDerivedCredentialv1'
    },
    117 => [
      VocabTermEncoder { termId: 108 },
      VocabTermEncoder { termId: 486 }
    ]
  }
}
```

Note that the original key of the undefined term `proofValue` had been preserved.

### CODE REVIEW

**Review by kdenhartog — DISMISSED** [review]
This one looks good to me

File: lib/Transformer.js [hunk 1]
```diff
@@ -369,7 +373,7 @@ export class Transformer {
   }
 
   _getTermType({activeCtx, def}) {
-    const {'@type': type} = def;
+    const {'@type': type} = def || {};
```
Comment (aliXsed) [inline lib/Transformer.js:414]: :+1:

**Review by aliXsed — DISMISSED** [review]
LGTM! with a good number of assuring tests.

**Review by ivy-wei-yang — DISMISSED** [review]

**Review by aliXsed — DISMISSED** [review]

**Review by ivy-wei-yang — DISMISSED** [review]

File: index.d.ts [hunk 2]
```diff
@@ -26,6 +26,10 @@ export type DiagnosticFunction = (message: string) => unknown;
  *   version 1, `0` to use no compression.
  * @param {Map} [options.appContextMap] - A map of JSON-LD Context URLs and
  *   their encoded CBOR-LD values (must be values greater than 32767 (0x7FFF)).
+ * @param {boolean} [options.compressionModeUndefinedTermAllowed] - Allow the
```
Comment (jhill-mattr) [inline index.d.ts:29]: Please mention what it gets defaulted to

Comment (zhenwenc) [reply index.d.ts:29]: Good point. Thanks! Will do 👍

**Review by ivy-wei-yang — APPROVED** [review]

**Review by aliXsed — APPROVED** [review]

**Pull request merged**